File: test/pytorch_test.py
import numpy as np
import torch
from torch import nn
from typing import List


class Mlp(nn.Module):
  def __init__(self, widths):
    super().__init__()
    self.layers = [] 
    for i in range(len(widths) - 1):
      if 0 < len(self.layers):
        self.layers.append(nn.ReLU())
      self.layers.append(nn.Linear(widths[i], widths[i + 1]))
      nn.init.xavier_uniform_(self.layers[-1].weight)
    # self.layers.append(nn.Softmax(dim=-1))
    self.layers = nn.Sequential(*self.layers)

  def forward(self, x):
    return self.layers(x)

def gen_input(dim: int, target: int):
  res = -np.random.random(dim).astype(np.float32) * 2
  res[target] = -1
  return res

def gen_label(dim: int, target: int):
  res = np.zeros(dim, dtype=np.float32)
  res[target] = 1
  return res

def validate(net: Mlp, dim: int, n: int):
  np.random.seed(22222)

  total_prob = 0.0
  for _ in range(n):
    target = np.random.randint(dim)
    in_tensor = torch.tensor(np.array([gen_input(dim, target)]))
    total_prob += nn.Softmax(dim=-1)(net.forward(in_tensor)).detach().numpy()[0][target]
  return total_prob / n

def run(widths: List[int], num_epochs: int, num_samples: int, batch_size: int, learning_rate: float):
  np.random.seed(2345)

  net = Mlp(widths)

  loss_function = nn.CrossEntropyLoss()
  optimizer = torch.optim.SGD(net.parameters(), lr=learning_rate, momentum=0.9, weight_decay=1e-4)

  for epoch in range(num_epochs):
    # validation
    print("epoch=%2d, avg_prob=%0.6f" % (epoch, validate(net, widths[0], num_samples)))
    
    # train
    for _ in range(num_samples // batch_size):
      optimizer.zero_grad()
      
      inputs = []
      labels = []
      for _ in range(batch_size):
        target = np.random.randint(widths[0])
        inputs.append(gen_input(widths[0], target))
        labels.append(gen_label(widths[0], target))
      inputs = np.array(inputs)
      labels = np.array(labels)

      outputs = net(torch.tensor(inputs))      
      loss = loss_function(outputs, torch.tensor(labels))
      loss.backward()
      optimizer.step()

  return net


if __name__ == '__main__':
  run(
    widths=[10, 50, 50, 10],
    num_epochs=100,
    num_samples=int(1e5),
    batch_size=32,
    learning_rate=3e-3,
  )

File: test/pytorch_test.cpp
#include "../nn.cpp"

#include <bits/stdc++.h>
#include <random>
#include <sys/time.h>
#include <utility>
#include <x86intrin.h>

#define rep(i, n) for (int i = 0; (i) < ((int) (n)); (i)++)
#define sz(v) ((int) ((v).size()))
#define all(v) (v).begin(), (v).end()
#define debug(v) { cerr << __LINE__ << ": " << (#v) << "=" << (v) << endl; }
#define debug2(v0, v1) { cerr << __LINE__ << ": " << (#v0) << "=" << (v0) << ", " << (#v1) << "=" << (v1) << endl; }
#define debug3(v0, v1, v2) { cerr << __LINE__ << ": " << (#v0) << "=" << (v0) << ", " << (#v1) << "=" << (v1) <<  ", " << (#v2) << "=" << (v2) << endl; }

using namespace std;
using namespace nn_eda;


float my_rand(float mini, float maxi)
{
  int mask = (1 << 25) - 1;
  float f = ((float) (rand() & mask)) / mask;
  return mini + (maxi - mini) * f;
}

vector<float> gen_input(int dim, int target)
{
  vector<float> res;
  rep(i, dim) res.push_back(-my_rand(0, 2));
  res[target] = -1;
  return res;
}

vector<float> gen_label(int dim, int target)
{
  vector<float> res(dim, 0);
  res[target] = 1.0;
  return res;
}

int get_target(int dim)
{
  return rand() % dim;
}

double validate_model(Nn &nn, int dim, int n)
{
  srand(3456);

  double total_prob = 0.0;
  rep(loop, n){
    int target = get_target(dim);
    vector<float> input = gen_input(dim, target);
    total_prob += nn.forward(input)[target];
  }
  double avg_prob = total_prob / n;
  return avg_prob;
}

Nn train_model(vector<int> widths, int num_epochs, int num_samples, int batch_size, double learning_rate)
{
  srand(2345);

  Nn nn(widths);

  rep(epoch, num_epochs){
    // validation
    {
      double avg_prob = validate_model(nn, widths[0], num_samples);
      debug2(epoch, avg_prob);
    }

    // train
    {
      rep(loop, num_samples / batch_size){
        vector<vector<float>> inputs;
        vector<vector<float>> labels;
        rep(i, batch_size){
          int target = get_target(widths[0]);
          inputs.push_back(gen_input(widths[0], target));
          labels.push_back(gen_label(widths[0], target));
        }
        nn.train(inputs, labels, learning_rate);
      }
    }
  }

  return nn;
}

int main()
{
  vector<int> widths({10, 50, 50, 10});
  int num_epochs = 100;
  int num_samples = (int) 1e5;
  int batch_size = 32;
  double learning_rate = 3e-3;
  train_model(widths, num_epochs, num_samples, batch_size, learning_rate);
}
